### ISSUE TITLE
Custom loss with extra argument in TF 2.0

### ISSUE THREAD

**Issue by lminer** [tensorflow/tensorflow#32142]
The traditional method of creating a custom loss function with an additional input for tf.keras no longer functions in tensorflow 2.0. It seems like the only way to do it now is with a custom training loop, which means you lose a lot of the convenience of keras (callbacks etc).

In the following case, the extra argument is the input data into the model, which is contained in a `Dataset`. In 1.14 case, I'd run `.make_one_shot_iterator().get_next()` on the dataset and then pass the tensor I get into the loss function. This isn't possible with eager execution, so instead I've tried to just pass the dataset.

```python
import numpy as np
import tensorflow as tf


class WeightedSDRLoss(tf.keras.losses.Loss):

    def __init__(self, noisy_signal, reduction=tf.keras.losses.Reduction.AUTO, name='WeightedSDRLoss'):
        super().__init__(reduction=reduction, name=name)
        self.noisy_signal = noisy_signal

    def sdr_loss(self, sig_true, sig_pred):
        return (-tf.reduce_mean(sig_true * sig_pred) /
                (tf.norm(tensor=sig_pred) * tf.norm(tensor=sig_true)))

    def call(self, y_true, y_pred):
        noise_true = self.noisy_signal - y_true
        noise_pred = self.noisy_signal - y_pred
        alpha = (tf.reduce_mean(tf.square(y_true)) /
                 tf.reduce_mean(tf.square(y_true) + tf.square(self.noisy_signal - y_pred)))
        return alpha * self.sdr_loss(y_true, y_pred) + (1 - alpha) * self.sdr_loss(noise_true, noise_pre$

data_x = np.random.rand(5, 4, 1)
data_y = np.random.rand(5, 4, 1)

x = tf.keras.layers.Input(shape=[4, 1])
y = tf.keras.layers.Activation('tanh')(x)
model = tf.keras.models.Model(inputs=x, outputs=y)

train_dataset = tf.data.Dataset.from_tensor_slices((data_x, data_y)).batch(1)

model.compile(loss=WeightedSDRLoss(x))
model.fit(train_dataset)
```

 I get the error:

````python
op_name = '__inference_distributed_function_169', num_outputs = 2
inputs = [<tf.Tensor: id=82, shape=(), dtype=resource, numpy=<unprintable>>, <tf.Tensor: id=83, shape=(), dtype=variant, numpy=<unprintable>>, <tf.Tensor 'input_1:0' shape=(None, 4, 1) dtype=float32>]
attrs = ('executor_type', '', 'config_proto', b'\n\x07\n\x03GPU\x10\x00\n\x07\n\x03CPU\x10\x012\x02J\x008\x01')
ctx = <tensorflow.python.eager.context.Context object at 0x11785f4e0>
name = None

    def quick_execute(op_name, num_outputs, inputs, attrs, ctx, name=None):
      """Execute a TensorFlow operation.
    
      Args:
        op_name: Name of the TensorFlow operation (see REGISTER_OP in C++ code) to
          execute.
        num_outputs: The number of outputs of the operation to fetch.
                     (Explicitly provided instead of being inferred for performance
                     reasons).
        inputs: A list of inputs to the operation. Each entry should be a Tensor, or
          a value which can be passed to the Tensor constructor to create one.
        attrs: A tuple with alternating string attr names and attr values for this
          operation.
        ctx: The value of context.context().
        name: Customized name for the operation.
    
      Returns:
        List of output Tensor objects. The list is empty if there are no outputs
    
      Raises:
        An exception on error.
      """
      device_name = ctx.device_name
      # pylint: disable=protected-access
      try:
        ctx.ensure_initialized()
        tensors = pywrap_tensorflow.TFE_Py_Execute(ctx._handle, device_name,
                                                   op_name, inputs, attrs,
>                                                  num_outputs)
E                                                  TypeError: An op outside of the function building code is being passed
E                                                  a "Graph" tensor. It is possible to have Graph tensors
E                                                  leak out of the function building context by including a
E                                                  tf.init_scope in your function building code.
E                                                  For example, the following function will fail:
E                                                    @tf.function
E                                                    def has_init_scope():
E                                                      my_constant = tf.constant(1.)
E                                                      with tf.init_scope():
E                                                        added = my_constant * 2
E                                                  The graph tensor has name: input_1:0

../../../anaconda3/envs/separate2/lib/python3.6/site-packages/tensorflow_core/python/eager/execute.py:61: TypeError

During handling of the above exception, another exception occurred:

    def test_loss():
    
        data_x = np.random.rand(5, 4, 1)
        data_y = np.random.rand(5, 4, 1)
    
        x = keras.layers.Input(shape=[4, 1])
        y = keras.layers.Activation('tanh')(x)
        model = keras.models.Model(inputs=x, outputs=y)
    
        train_dataset = tf.data.Dataset.from_tensor_slices((data_x, data_y)).batch(1)
        print(train_dataset)
    
        model.compile(loss=WeightedSDRLoss(x))
>       model.fit(train_dataset)

test_preprocess.py:162: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../anaconda3/envs/separate2/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training.py:734: in fit
    use_multiprocessing=use_multiprocessing)
../../../anaconda3/envs/separate2/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2.py:324: in fit
    total_epochs=epochs)
../../../anaconda3/envs/separate2/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2.py:123: in run_one_epoch
    batch_outs = execution_function(iterator)
../../../anaconda3/envs/separate2/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2_utils.py:86: in execution_function
    distributed_function(input_fn))
../../../anaconda3/envs/separate2/lib/python3.6/site-packages/tensorflow_core/python/eager/def_function.py:445: in __call__
    return self._concrete_stateful_fn._filtered_call(canon_args, canon_kwds)  # pylint: disable=protected-access
../../../anaconda3/envs/separate2/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py:1141: in _filtered_call
    self.captured_inputs)
../../../anaconda3/envs/separate2/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py:1224: in _call_flat
    ctx, args, cancellation_manager=cancellation_manager)
../../../anaconda3/envs/separate2/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py:511: in call
    ctx=ctx)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

op_name = '__inference_distributed_function_169', num_outputs = 2
inputs = [<tf.Tensor: id=82, shape=(), dtype=resource, numpy=<unprintable>>, <tf.Tensor: id=83, shape=(), dtype=variant, numpy=<unprintable>>, <tf.Tensor 'input_1:0' shape=(None, 4, 1) dtype=float32>]
attrs = ('executor_type', '', 'config_proto', b'\n\x07\n\x03GPU\x10\x00\n\x07\n\x03CPU\x10\x012\x02J\x008\x01')
ctx = <tensorflow.python.eager.context.Context object at 0x11785f4e0>
name = None

    def quick_execute(op_name, num_outputs, inputs, attrs, ctx, name=None):
      """Execute a TensorFlow operation.
    
      Args:
        op_name: Name of the TensorFlow operation (see REGISTER_OP in C++ code) to
          execute.
        num_outputs: The number of outputs of the operation to fetch.
                     (Explicitly provided instead of being inferred for performance
                     reasons).
        inputs: A list of inputs to the operation. Each entry should be a Tensor, or
          a value which can be passed to the Tensor constructor to create one.
        attrs: A tuple with alternating string attr names and attr values for this
          operation.
        ctx: The value of context.context().
        name: Customized name for the operation.
    
      Returns:
        List of output Tensor objects. The list is empty if there are no outputs
    
      Raises:
        An exception on error.
      """
      device_name = ctx.device_name
      # pylint: disable=protected-access
      try:
        ctx.ensure_initialized()
        tensors = pywrap_tensorflow.TFE_Py_Execute(ctx._handle, device_name,
                                                   op_name, inputs, attrs,
                                                   num_outputs)
      except core._NotOkStatusException as e:
        if name is not None:
          message = e.message + " name: " + name
        else:
          message = e.message
        six.raise_from(core._status_to_exception(e.code, message), None)
      except TypeError as e:
        keras_symbolic_tensors = [
            x for x in inputs if ops._is_keras_symbolic_tensor(x)
        ]
        if keras_symbolic_tensors:
          raise core._SymbolicException(
              "Inputs to eager execution function cannot be Keras symbolic "
>             "tensors, but found {}".format(keras_symbolic_tensors))
E         tensorflow.python.eager.core._SymbolicException: Inputs to eager execution function cannot be Keras symbolic tensors, but found [<tf.Tensor 'input_1:0' shape=(None, 4, 1) dtype=float32>]
````

Basically seems like the previous way of passing extra arguments into a loss function in keras is broken.

I'm running tensorflow 2.0-rc0 and this happens regardless of the platform.

**Comment by oanush** [comment]
@lminer ,
In order to expedite the trouble-shooting process, please provide complete code snippet to reproduce the issue reported here. Thanks!

**Comment by lminer** [comment]
@oanush I updated it with running code.

**Comment by lminer** [comment]
I tested this and it seems like it works in the beta version.

**Comment by bersbersbers** [comment]
@lminer may I ask which of the two codes above you tested, and in which "beta" version? Is that rc0, rc1, or nightly?

**Comment by ymodak** [comment]
The provided code snippet looks incomplete. Can you please update it? Also can you please test with tf 2.0 nightly version and check if issue still persists?
```pip install tf-nightly-2.0-preview```
Thanks!

**Comment by lminer** [comment]
@ymodak I can confirm that the error is still present on the nightly preview. I updated the code so it should run no problem now.

**Comment by lminer** [comment]
@pavithrasv, @ymodak As an update, this still does not work in the stable release of Tensorflow 2.0. However, if I set the flag `experimental_run_tf_function=False` in `model.compile` it runs without a problem.

At the end of the day, this is a bit of a hack in order to pass the model input into the cost function. It might be worthwhile to simply include a hook for model input in `keras.losses.Loss`. There are plenty of custom loss functions that require this input.

**Comment by feature-engineer** [comment]
@lminer Thanks for the `experimental_run_tf_function=False` tip, what exactly does it do? It's not documented anywhere...

**Comment by feature-engineer** [comment]
@lminer Is this a good solution for you?:

```python
import numpy as np
import tensorflow as tf


class WeightedSDRLoss(tf.keras.losses.Loss):

    def __init__(self, noisy_signal, reduction=tf.keras.losses.Reduction.AUTO, name='WeightedSDRLoss'):
        super().__init__(reduction=reduction, name=name)
        self.noisy_signal = noisy_signal

    def sdr_loss(self, sig_true, sig_pred):
        return (-tf.reduce_mean(sig_true * sig_pred) /
                (tf.norm(tensor=sig_pred) * tf.norm(tensor=sig_true)))

    def call(self, y_true, y_pred):
        noise_true = self.noisy_signal - y_true
        noise_pred = self.noisy_signal - y_pred
        alpha = (tf.reduce_mean(tf.square(y_true)) /
                 tf.reduce_mean(tf.square(y_true) + tf.square(self.noisy_signal - y_pred)))
        return alpha * self.sdr_loss(y_true, y_pred) + (1 - alpha) * self.sdr_loss(noise_true, noise_pred)

data_x = np.random.rand(5, 4, 1)
data_y = np.random.rand(5, 4, 1)

y_true = tf.keras.layers.Input(shape=[4, 1])
x = tf.keras.layers.Input(shape=[4, 1])
y_pred = tf.keras.layers.Activation('tanh')(x)
model = tf.keras.models.Model(inputs=[x, y_true], outputs=y_pred)
model.add_loss(WeightedSDRLoss(x)(y_true, y_pred))

train_dataset = tf.data.Dataset.from_tensor_slices(((data_x, data_y),)).batch(1)

model.compile()
model.fit(train_dataset)
```

**Comment by lminer** [comment]
@feature-engineer, it's definitely better, but the other problem is loading a serialized version of the model. I have to do it in a seriously hacky way right now, where I recompile the model all over again with the same loss function. I use the original optimizer so I can preserve that state.

**Comment by dgrahn** [comment]
I encountered a problem with the same solution.
I can't say that it was the *same* problem, but setting `experimental_run_tf_function = False` on the compile method solved it.

I'll update with the error message as soon as I am able to reproduce it, but it was essentially.
```
Tensor cannot be converted to numpy array.
```

**Comment by wmcnally** [comment]
@lminer What should I do if the argument I want to pass to the loss function is not the input to the model but a different tensor in my Dataset?

**Comment by yourtheron** [comment]
@lminer @ymodak 
I'd like to report a follow-up issue. I tried to run Lminer's code snippet on TF2.0.0--I use Spyder 3.3.6, Python 3.7 I got an error
`
	TypeError: Cannot convert 'auto' to EagerTensor of dtype float
`
I also tried to build up my own custom loss with the same structure as lminer's, the same issue was raised.
Is that another bug of tf?

Plus, the docs 
https://www.tensorflow.org/api_docs/python/tf/keras/losses/Reduction?version=stable
is a bit confusing. If my minibatch size is larger than 1, and if I use 'AUTO' option, shall I average my final loss over the minibatch? Do both 'AUTO' and 'SUM_OVER_BATCH_SIZE' mean that I will NOT need to do so if my code has nothing to do with distributed strategy?

**Comment by founta** [comment]
This workaround using `experimental_run_tf_function = False` in the compile line does not appear to work as of the newest Tensorflow version, 2.2.0, with Python 3.8 but does work fine with Tensorflow version 2.1.0 with Python 3.7. Both with Keras 2.3.1.

**Comment by jvishnuvardhan** [comment]
@lminer Is this still an issue for you. I tried `TF2.4.1` and was able to save and load and didn't face any issue. [Here](https://colab.research.google.com/gist/jvishnuvardhan/e709aa673d747c9adba48b9efdadf9d3/untitled.ipynb) is a gist for reference. 

Please close the issue if this was resolved for you. Thanks!

If anyone has any similar issue, please share a standalone code to reproduce the issue. If you have any new issue, please file a new issue. thanks!

**Comment by lminer** [comment]
Looks good
